## Valores e Tipos

* O valor constitui uma das manipulações fundamentais do programa
    - 2, resultado de 1+1
    - "Olá, mundo!"
* Valores pertencem a tipos diferentes:
    - 2 representa o tipo inteiro
    - 2.3 representa o tipo real (float)
    - "Olá, mundo!" representa uma cadeia de caractéres (string)
    - True, representa um valor booleano

## Variáveis

Na matemática uma variável pode ser entendido como um símbolo. Isto representa uma característica (quantidade, qualidade, magnitude etc) que pode possuir valores numéricos.

* \begin{equation}f(x) = x+1\end{equation}
    - A variável x exerce o papel de parâmetro de entrada da função 
* \begin{equation}x^2 - x + 1 = 0\end{equation}
    - A variável x exerce o papel de incógnita
    
Para a Computação, uma variável se refere a uma posição em memória capaz de reter um valor ou expressão. Por meio de uma analogia, pode-se considerar a variável como uma caixa e o valor seu conteúdo.

A criação de uma variável utiliza o comando de atribuição ``=``, além disto têm-se algumas regras. Os nomes de variável podem conter letras, números e o caractére ``_``, porém não podem ser iniciadas por números.
* Escolha nomes significativos para as variáveis;
* Os nomes podem ser arbitrariamente longos;
* Os nomes são sensíveis a maiúsculas e minúsculas, logo ``nome`` será diferente de ``Nome``.

In [13]:
mensagem = "Sou da UnB"
idade = 18
pi = 3.1415
verdade = True

In [16]:
1tamanho = 4

SyntaxError: invalid syntax (<ipython-input-16-2479865f6f98>, line 1)

Além destas regras, no Python existem algumas palavras reservadas. Esta lista pode ser obtida por meio de dois comandos:

In [25]:
import keyword
print(keyword.kwlist)

['False', 'None', 'True', 'and', 'as', 'assert', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']


Uma maneira de representar as variáveis em papel se dá por diagramos de estado. Este diagrama contém o nome da variável e uma seta apontando para seu estado (conteúdo) atual.

<img src = "./img/diagrama_estado.png">

No python a função ``print`` fica responsável por mostrar o conteúdo de variáveis ou expressões na tela.

In [15]:
print(mensagem)
print(idade)
print(pi)
print(verdade)

Sou da UnB
18
3.1415
True


Note que previamente foi mencionado que os valores possuem tipo, logo as variáveis possuem o mesmo tipo do valor de seu conteúdo. Este tipo pode ser acessado pela função ``type``.

In [14]:
print(type(mensagem))
print(type(idade))
print(type(pi))
print(type(verdade))

<class 'str'>
<class 'int'>
<class 'float'>
<class 'bool'>


Uma variável pode exercer diversos papéis no programa, tais como os vistos na tabela abaixo.

 Papel | Descrição 
-------|------------
Constante | Recebe um valor apenas uma vez e não o muda
Contador  | Recebe valores sucessivos de forma sistemática
Somador   | Guarda a soma do seu valor atual mais alguma quantidade
Sinalizador        | Sinaliza algum ocorrido
Valor mais recente | Guarda o último valor
Melhor valor       | Guarda o valor mais apropriado

Entre estes estão os mais utilizados, porém também podem exercer outros papéis que serão explicados conforme a necessidade.

### Exemplo
Suponha um aluno com insônia. Um método que pode ser usado para dormir é o de contar carneiros.

<img src="./img/insonia.png">

In [21]:
ovelhas = 0
acordado = True
from random import randint
dormir = randint(0, 20)
while acordado:
    print(f"Ovelhas = {ovelhas}")
    ovelhas += 1
    if ovelhas > dormir:
        acordado = False
print("Dormiu")

Ovelhas = 0
Ovelhas = 1
Ovelhas = 2
Ovelhas = 3
Ovelhas = 4
Ovelhas = 5
Ovelhas = 6
Ovelhas = 7
Ovelhas = 8
Ovelhas = 9
Ovelhas = 10
Ovelhas = 11
Ovelhas = 12
Ovelhas = 13
Ovelhas = 14
Ovelhas = 15
Dormiu


De acordo com ``Abelson et al. 1996``.
* Um programa de computador não é apenas uma lista de comandos e variáveis em uma linguagem de programação para algum computador o executar;
* Ele é mais que isso, é um novo meio para expressar ideias sobre metodologias;
* Portanto, programas tem que ser escritos para as pessoas lerem, e apenas, por acaso, para serem executados por máquinas.

### Exemplo 2
Tente adivinhar o que o seguinte programa faz:

In [27]:
e = 3.1415
n = 10
z = e * n**2
print(z)

314.15000000000003


Com esta escolha de nomeação torna-se uma tarefa muito complicada adivinhar o que o programa esta calculando. Porém com uma nomeação correta,

In [28]:
pi = 3.1415
raio = 10
area = pi * raio**2
print(area)

314.15000000000003


siga o conselho, escreva programas para seres humanos lerem, que por acaso, possam ser executados por máquinas.

## Expressões

Uma expressão representa uma combinação de valores, variáveis e operadores. Se for digitado uma expressão na linha de comando, o interpretador avalia e exibe o resultado.

In [29]:
1+1

2

In [32]:
x = 1+1

Embora expressão contenham valores, variáveis e operadores, nem toda expressão contém todos estes elementos. Um valor po si só é considerado uma expressão, do mesmo modo que uma variável

In [30]:
17

17

In [33]:
x

2

A avaliação de uma expressão não é exatamente a mesma coisa que imprimir o valor

In [34]:
mensagem = "E aí, aluno?"
mensagem

'E aí, aluno?'

In [35]:
print(mensagem)

E aí, aluno?


Quando o Python exibe o valor de uma expressão, usa o mesmo formato para a atribuição. No caso de ``strings``, isso significa que as aspas estão incluídas. Mas o comando ``print`` imprime o valor da expressão, que, neste caso, é o conteúdo da string sem as aspas.

Em um ``script`` uma expressão sozinha é válida, porém não tem efeito.

In [38]:
17
3.2
"Olá"
1+1
print("ok")

ok


A única saída do script acima será o que está dentro do comando print. Você poderia mudar os valores para exibir todas as expressões?

## Operadores e operandos
Operadores são símbolos especiais que representam computações como adição e multiplicação. Os valores que o operador usa são chamados de operandos, todas as expressões seguintes são válidas, em Python, e seus significados são similares a matemática.

In [1]:
print(20+32)
print(5**2)
minutos = 237
horas = minutos/60
print(horas, minutos//60)
print((5+9)*(15-7))


52
25
3.95 3
112


Em Python, os símbolos `+`, `-`, `/` e o uso de parênteses tem equivalência com seus significados matemáticos. O asterisco `*` representa uma multiplicação, e `**` uma potencição. O `//` significa uma divisão com resultado, somente, inteiro. Quando o nome de uma variável aparece no lugar de um valor, ela será substituída pelo seu conteúdo antes da operação ser executada.

Note que a divisão inteira retorna um valor inteiro, logo:

In [3]:
minutos = 59
print(minutos/60)
print(minutos//60)

0.9833333333333333
0


59 minutos divido por 60, resultam em 0.983 horas, porém ao usar a divisão inteira resultam em 0.

### Conversão de tipos

função | cast
-------|--------
int()  | inteiro
float()| real
str()  | string
bool() | booleano

Pode-se convertem os valores entre os tipos existentes.

In [4]:
print(int(3.14))
print(int(3.9999))
print(int(3.0))
print(int("2345"))
print(float(17))
print(float("17.25"))
print(str(17))
print(str(123.45))

3
3
3
2345
17.0
17.25
17
123.45


Note que int(3.9999) não arredonda para o inteiro mais próximo, funcionando como chão. A função `round` irá fazer o arredondamento.

In [5]:
print(round(3.9999))

4


Quando mais de uma operação aparece em uma mesma expressão deve-se levar em conta uma ordem de precedência. Python possui as mesmas regras da matemática. O acrônimo `PEMDAS` é uma maneira de lembrar a ordem destas operações.
* __P__ Parênteses têm a mais alta precedência e podem ser usados para forçar uma expressão a ser avaliada na ordem necessária;
* __E__ Exponenciação ou potênciação, constitui a próxima operação na fila de precedência.;
* __MD__ Multiplicação e Divisão possuem a mesma precedência e são as terceiras a serem executadas;
* __AS__ Por fim, adição e subtração são as últimas a serem resolvidas.

Operadores com a mesma precedência são avaliados da esquerda para a direita. Por algum motivo o Python não segue esta regra para a potênciação. Assim:

In [10]:
print(2**3**2)
print((2**3)**2) # Note a diferença que fazem os parênteses

512
64


## Operações com strings

De maneira geral, não se pode executar operações matemáticas em strings, ainda que as strings sejam números. O que segue é inválido.

In [11]:
mensagem = "Olá"
print(mensagem-1)

TypeError: unsupported operand type(s) for -: 'str' and 'int'

In [13]:
print(mensagem*"teste")

TypeError: can't multiply sequence by non-int of type 'str'

In [14]:
print("15"+2)

TypeError: must be str, not int

Porém os operadores matemáticos possuem operações específicas para as strings. Por exemplo `+` significa concatenação, juntar duas sequências.

In [15]:
fruta = "banana"
assada = "com canela"
print(fruta + assada)

bananacom canela


O `*` serve para repetir a mesma cadeia arbitrariamente.

In [16]:
print("legal"*3)

legallegallegal


Existem diversos outros métodos referentes a strings, porém este tópico será revisitado mais a frente no curso.

## Entrada de dados

Para receber dados do usuário o desenvolvedor pode utilizar a função `input`, por exemplo:

In [18]:
nome = input("Digite seu nome: ")
print(nome)

Digite seu nome: Caetano
Caetano


O usuário digita seu nome e aperta ENTER em seguida, o texto digitado será do tipo `str` e a variável nome irá recebê-lo. Caso o objetivo seja ler um número deve-se lembrar de fazer a conversar do tipo. Além disso, em algumas vezes pode ser requerido ler mais de um valor por linha, para tal a cadeia deve ser quebrada por algum tipo de marcação. Para o exemplo a marcação será o caractére ESPAÇO.

In [21]:
x, y = input().split(' ')
x = int(x)
y = int(y)
print(x, y)

1 2
1 2


## Composição

In [22]:
resposta = input("Qual o raio?")
r = float(resposta)
area = 3.14159 * r**2
print(f'A área é: {area}')

Qual o raio?2
A área é: 12.56636


O exempo acima não utiliza a composição, seu uso recai na mesclagem de várias funções e comandos numa mesma linha. Isto pode tornar o código mais conciso, porém também torna-o mais complicado de se ler.

In [24]:
r = float(input("Qual é o raio?"))
area = 3.14159 * r**2
print(f'A área é: {area}')

Qual é o raio?2
A área é: 12.56636


In [26]:
print(f"A área é: {3.14159 * float(input('Qual é o raio?'))**2}")

Qual é o raio?2
A área é: 12.56636


### Operação módulo
A operação módulo (MOD) constitui uma operação matemática que busca o resto de uma divisão. Em Python o operador `%` é o símbolo para tal operação.

In [27]:
q = 7 // 3 # divisão inteira
print(q)
r = 7 % 3 # módulo
print(r)

2
1


Esta operação é extramente útil, serve, por exemplo, para verificar se um número é divisível por outro. Assim,
* Se x % y == 0, então y divide x
    * Como descobrir se um número é par? E ímpar?

Esta operação também pode ser útil para fazer conversões, por exemplo segundos para dias, horas, minutos e segundos.

In [31]:
milhao_segundos = 10e5
horas = milhao_segundos // 3600
segundos_restantes = milhao_segundos % 3600
minutos = segundos_restantes // 60
segundos_finais = segundos_restantes % 60
dias = horas // 24
horas_restantes = horas % 24
print(f"Dias = {dias}, horas = {horas_restantes}, minutos = {minutos}, segundos = {segundos_finais}")

Dias = 11.0, horas = 13.0, minutos = 46.0, segundos = 40.0


In [34]:
bilhao_segundos = 10e8
horas = bilhao_segundos // 3600
segundos = bilhao_segundos % 3600
minutos = segundos // 60
segundos = segundos % 60
dias = horas // 24
horas = horas % 24
anos = dias // 365
dias = dias % 365
meses = dias // 30
dias = dias % 30
print(f'Anos = {anos}, meses = {meses}, dias = {dias}, horas = {horas}, minutos = {minutos}, segundos = {segundos}')

Anos = 31.0, meses = 8.0, dias = 19.0, horas = 1.0, minutos = 46.0, segundos = 40.0


## Exemplo Média Aritmética

* Enunciado
    * Leia dois números inteiros, calcule a média aritmética e mostre o resultado
* Entrada
    * 2 números inteiros por linha
* Saída
    * Média aritmética da linha

Exemplo Entrada | Exemplo Saída
----------------|---------------
1 1 | 1.0
1 2 | 1.5

In [35]:
a, b = input().split()
a, b = int(a), int(b)
media = (a+b)/2
print(media)

1 2
1.5


## Bibliografia
* How to Think like a computer scientist: Learning with Python 3 - Documentation - Release 3rd Edition
    * Peter Wentworth, Jeffrey Elkner, Allen B. Downey and Chris Meyers - Apr 26, 2017